x-pack/platform/plugins/shared/fleet/server/services/setup.ts

```diff
@@ -78,6 +80,20 @@ export interface SetupStatus {
   >;
 }
 
+export async function _runSetupWithLock(setupFn: () => Promise<SetupStatus>) {
+  return await pRetry(
+    () => appContextService.getLockManagerService()!.withLock('fleet-setup', () => setupFn()),
+    {
+      onFailedAttempt: async (error) => {
+        if (!(error instanceof LockAcquisitionError)) {
+          throw error;
+        }
+      },
+      maxRetryTime: 5 * 60 * 1000, // Retry for 5 minute to get the lock
+    }
+  );
+}
```

Comment:
I think there is a bug in the retry logic here. From the `pRetry` documentation:

> If the `onFailedAttempt` function throws, all retries will be aborted and the original promise will reject with the thrown error.

Right now all errors but `LockAcquisitionError` are thrown, meaning they are aborted without retries. Only LockAcquisitionError's are retried.

Is that the intention?

Reply:
Yes that the intention, we already have some retry logic in the Fleet setup.

Reply:
All right. Why would you want to retry `setupFn`, if it's already running? Then it would run multiple times - although sequentially, not in parallel.

Reply:
> All right. Why would you want to retry setupFn, if it's already running? Then

We want to be sure the setup ran and finish  at least once If it's running on a different instance Kibana, but that instance is killed for any reason, this way we will retry, if the setup was already successful that operation should be relatively quick.